### PR TITLE
Tchannel alpha4 DO NOT MERGE

### PR DESCRIPTION
**Pull request by Raynos**
Opening PR to be able to leave comments. 

This wont work without `tchannel#bug-fix-thing` branch.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -265,57 +266,69 @@ RingPop.prototype.bootstrap = function bootstrap(opts, callback) {
         return;
     }
 
-    var start = new Date();
-
-    this.seedBootstrapHosts(bootstrapFile);
+    if (this.channel && this.channel.address() === null) {
+        this.channel.once('listening', doRealBootstrap);
```
Comment (Raynos) [inline index.js]: tchannel is now async open.

You cannot create it and send a request to it.

You have to wait for listening.

This makes bootstrap even more async.

File: lib/join_cluster.js [hunk 2]
```diff
@@ -338,7 +338,8 @@ JoinCluster.prototype.joinNode = function joinNode(node, callback) {
     var self = this;
     var joinOpts = {
         host: node,
-        timeout: this.joinTimeout
+        timeout: this.joinTimeout,
+        service: 'ringpop'
```
Comment (Raynos) [inline lib/join_cluster.js]: we must send a service === ringpop with every outgoing call request.

File: lib/join_cluster.js [hunk 3]
```diff
@@ -346,19 +347,27 @@ JoinCluster.prototype.joinNode = function joinNode(node, callback) {
         incarnationNumber: this.ringpop.membership.localMember.incarnationNumber
     });
 
-    this.ringpop.channel.send(joinOpts, '/protocol/join', null, joinBody, function onSend(err, head, body) {
-        if (err) {
-            return callback(err, node);
-        }
+    console.log('how many calls ?!?!?!');
+    this.ringpop.channel
```
Comment (Raynos) [inline lib/join_cluster.js]: interface changed to `.send().request()`

The callback changed to `(err, resp)` where `resp.ok` is true/false (like statusCode === 200).

File: lib/join_cluster.js [hunk 4]
```diff
@@ -346,19 +347,27 @@ JoinCluster.prototype.joinNode = function joinNode(node, callback) {
...
+        .request(joinOpts)
+        .send('/protocol/join', null, joinBody, function onSend(err, res) {
+            console.log('wtf', err, res);
+            if (err) {
+                return callback(err, node);
+            }
 
-        var bodyObj = safeParse(body.toString());
+            if (!res.ok) {
+                return callback(new Error(String(res.arg3)), node);
```
Comment (Raynos) [inline lib/join_cluster.js]: In the server for ringpop helper we put the message in arg3.

File: lib/request-proxy/index.js [hunk 5]
```diff
@@ -116,7 +116,14 @@ proto.proxyReq = function proxyReq(opts) {
                 max: numOrDefault(maxRetries, self.maxRetries),
                 schedule: retrySchedule || self.retrySchedule
             }
-        }, onProxy);
+        }, function () {
+            var args = arguments;
+            var self = this;
+
+            process.nextTick(function () {
```
Comment (Raynos) [inline lib/request-proxy/index.js]: nextTick because hack reasons;

File: lib/tchannel-alpha4.js [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+'use strict';
+
+var recvPingReq = require('./swim/ping-req-recvr.js');
```
Comment (Raynos) [inline lib/tchannel-alpha4.js]: This file is used with tchannel 2 instead of `.setupTChannel()` method.

File: lib/tchannel-alpha4.js [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+
+    '/proxy/req': 'proxyReq'
+};
+
+function RingPopTChannel(ringpop) {
+    this.ringpop = ringpop;
+
+    var self = this;
+
+    return function handleRequest(req, res) {
```
Comment (Raynos) [inline lib/tchannel-alpha4.js]: The interface is now req/res instead of the old one.

File: lib/tchannel-alpha4.js [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+        if (!fn) {
+            return false;
+        }
+
+        fn.call(self, req.arg2, req.arg3, req.remoteAddr, onResponse);
+        return true;
+
+        function onResponse(err, res1, res2) {
+            if (err) {
+                res.sendNotOk(null, err.message);
```
Comment (Raynos) [inline lib/tchannel-alpha4.js]: sendNotOk() will do a call response where `resp.ok === false`. We put the message as a string in arg3

File: test/integration/join-test.js [hunk 9]
```diff
@@ -24,99 +24,99 @@ var test = require('tape');
...
-}, 'mega cluster', function t(bootRes, cluster, assert) {
-    assert.equal(cluster.length, 25, 'cluser of 25');
-
-    cluster.forEach(function eachNode(node) {
-        assert.ok(node.isReady, 'node is bootstrapped');
-    });
-
-    assert.end();
-});
+// test('bootstrap with self is ok', function t(assert) {
```
Comment (Raynos) [inline test/integration/join-test.js]: forgot to uncomment lol.

File: test/integration/proxy-test.js [hunk 10]
```diff
@@ -19,6 +19,10 @@
 // THE SOFTWARE.
 'use strict';
 
+require('leaked-handles').set({
```
Comment (Raynos) [inline test/integration/proxy-test.js]: forgot to remove this lol

File: lib/join_cluster.js (hunk 2 above)
Comment (jwolski) [inline lib/join_cluster.js]: kewl. i'll eventually want to centralize this logic so that a call site doesnt "forget"

**Pull request closed without merging**